From 752d559e63236f21bded0dc31e1d3763cbb6833b Mon Sep 17 00:00:00 2001
From: Andre Klapper <a9016009@gmx.de>
Date: Fri, 23 Aug 2024 13:14:23 +0200
Subject: [PATCH 04/14] Fix PHP 8.1 "preg_match(null)" exception in XHPASTNode

Summary:
When `$u` is `null`, `null` is passed to `preg_match()` which is deprecated behavior since PHP 8.1.
Thus first check if `$u === null`.

```
ERROR 8192: preg_match(): Passing null to parameter #2 ($subject) of type string is deprecated at [/var/www/html/phorge/arcanist/src/parser/xhpast/api/XHPASTNode.php:243]
```

Refs T15926

Test Plan: Either read the code, or probably run something like `arc unit` on D25797?

Reviewers: O1 Blessed Committers, valerio.bozzolan

Reviewed By: O1 Blessed Committers, valerio.bozzolan

Subscribers: tobiaswiese, valerio.bozzolan, Matthew, Cigaryno

Maniphest Tasks: T15926

Differential Revision: https://we.phorge.it/D25801
---
 src/parser/xhpast/api/XHPASTNode.php | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/parser/xhpast/api/XHPASTNode.php b/src/parser/xhpast/api/XHPASTNode.php
index 94c095dd3b3a..99adbe623c9e 100644
--- a/src/parser/xhpast/api/XHPASTNode.php
+++ b/src/parser/xhpast/api/XHPASTNode.php
@@ -240,7 +240,7 @@ final class XHPASTNode extends AASTNode {
         switch ($c) {
           case 'x':
             $u = isset($value[$ii + 1]) ? $value[$ii + 1] : null;
-            if (!preg_match('/^[a-f0-9]/i', $u)) {
+            if ($u === null || !preg_match('/^[a-f0-9]/i', $u)) {
               // PHP treats \x followed by anything which is not a hex digit
               // as a literal \x.
               $out .= '\\\\'.$c;
-- 
2.48.1

